metricbeat/module/mongodb/replstatus/info.go

```diff
@@ -43,85 +43,109 @@ type CollSize struct {
+	// 		now: 'Wed Jan 08 2025 11:45:14 GMT+0530 (India Standard Time)'
+	// 	}
+	// i.e., we get tFirst and tLast from oplog.rs which is the first and last
+	// timestamp of the oplog.
+	// Source from the same is here:
+	// 	https://github.com/mongodb/mongo/blob/20cbee37a0ee4d40b35d08b6a34ade81252f86a8/src/mongo/shell/db.js#L863
+	// This is how they calculate tFirst and tLast:
+	// 	https://github.com/mongodb/mongo/blob/20cbee37a0ee4d40b35d08b6a34ade81252f86a8/src/mongo/shell/db.js#L889
+	// So ideally, we will replicate the same logic here:
+	// 	var firstc = ol.find().sort({$natural: 1}).limit(1);
+	// 	var lastc = ol.find().sort({$natural: -1}).limit(1);
+	//
+	// oplog.rs is designed to scanned in natural ($natural) order. So, when
+	// querying without any sort, it will return the first entry in natural order.
+	// When we sort in reverse natural order (i.e., $natural: -1), it will return
+	// the last entry in natural order.
+	//
+	// The queries use optimizations like FindOne() and SetProjection() to limit
+	// the amount of data that needs to be scanned and processed.
+	// Projection is used to limit the fields returned in the query results i.e., we
+	// only need the timestamp (ts) field. FindOne() is used to retrieve a single
+	// document from the collection (limit: 1).
+
+	ctx := context.TODO()
+
+	var firstDoc struct {
+		Timestamp time.Time `bson:"ts"`
 	}
 
-	cursor, err := collection.Aggregate(context.Background(), pipeline)
+	// Get oldest entry using natural order
+	firstOpts := options.
+		FindOne().
+		SetProjection(bson.D{{Key: "ts", Value: 1}})
+	err := collection.FindOne(ctx, bson.D{}, firstOpts).Decode(&firstDoc)
 	if err != nil {
-		return 0, 0, fmt.Errorf("could not get operation timestamps in op log: %w", err)
+		return 0, 0, fmt.Errorf("first timestamp query failed: %w", err)
 	}
-	defer cursor.Close(context.Background())
 
-	var result struct {
-		MinTS time.Time `bson:"minTS"`
-		MaxTS time.Time `bson:"maxTS"`
+	// Get newest entry using reverse natural order
+	var lastDoc struct {
+		Timestamp time.Time `bson:"ts"`
 	}
-
-	if !cursor.Next(context.Background()) {
-		return 0, 0, errors.New("no documents found in op log")
-	}
-	if err := cursor.Decode(&result); err != nil {
-		return 0, 0, fmt.Errorf("error decoding response for timestamps: %w", err)
+	lastOpts := options.
+		FindOne().
+		SetProjection(bson.D{{Key: "ts", Value: 1}}).
+		SetSort(bson.D{{Key: "$natural", Value: -1}})
+	err = collection.FindOne(ctx, bson.D{}, lastOpts).Decode(&lastDoc)
+	if err != nil {
+		return 0, 0, fmt.Errorf("last timestamp query failed: %w", err)
 	}
 
-	minTS := uint32(result.MinTS.Unix())
-	maxTS := uint32(result.MaxTS.Unix())
-
-	return minTS, maxTS, nil
-}
-
-func contains(s []string, x string) bool {
-	for _, n := range s {
-		if x == n {
-			return true
-		}
-	}
-	return false
+	return uint32(firstDoc.Timestamp.Unix()), uint32(lastDoc.Timestamp.Unix()), nil
```

Comment:
Is it ok if we convert it to uint32?
Hoping there will be no overflow issues since we are converting Unix() which is an int64 to uint32.

Reply:
So, it works upto year 2038 then it can overflow. But yes kept it like to not change the types used previously. But makes sense, let's make this correct.